Documentation/design-docs/hw-intrinsics.md

```diff
@@ -0,0 +1,81 @@
+Implementation of Hardware Intrinsics in CoreCLR
+================================================
+This document describes the implementation of hardware intrinsics in CoreCLR.
+For information about how the intrinsic APIs are designed, proposed and approved,
+see https://github.com/dotnet/designs/blob/master/accepted/platform-intrinsics.md.
+
+In discussing the hardware intrinsics, we refer to the target platform, such as X86 or Arm64, as the "platform" and each set of extensions that are implemented as a unit (e.g. AVX2 on X64 or Simd on Arm64) as an "ISA".
+
+There is a design document for the Arm64 intrinsics: https://github.com/dotnet/coreclr/blob/master/Documentation/design-docs/arm64-intrinsics.md. It should be updated to reflect current (and ongoing) progress.
+
+## Overview
+
+The reference assemblies for the hardware intrinsics live in corefx, but all of the implementation is in the coreclr repo:
+
+* The C# implementation lives in coreclr/src/shared/System/Runtime/Intrinsics. These are little more than skeleton methods that usually invoke themselves recursively (see Indirect invocation support below).
+
+## C# Implementation
+
+The hardware intrinsics operate on and produce both primitive types (`int`, `float`, etc.) as well as vector types. The vector types are considered platform-agnostic, though not all platforms define intrinsics on all of these types. They are:
+
+* `Vector64<T>` - A 64-bit vector of type `T`. For example, a `Vector64<int>` would hold two 32-bit integers.
+* `Vector128<T>` - A 128-bit vector of type `T`
+* `Vector256<T>` - A 256-bit vector of type `T`
+
+Note that these are generic types, which distinguishes these from native intrinsic vector types. It also somewhat complicates interop, as the runtime currently doesn't support interop for generic types.
+*** Is there an issue for this?? ***
+
+The C# declaration of a hardware intrinsic ISA class is marked with the `[Intrinsic]` attribute, and the implementations of the intrinsic methods on that class are recursive. When the VM encounters such a method, it will communicate to the JIT that this is an intrinsic method, and will also pass a `mustExpand` flag to indicate that the JIT must generate code.
+
+### Platform-agnostic vector types
+
+The vector types supported by one or more target ISAs are supported across platforms, though they extent to which operations on them are available and accelerated is dependent on the target ISA. These are the generic types: `Vector64<T>`, `Vector128<T>` and `Vector256<T>`.
+
+## JIT Implementation
+
+The bulk of the implementation work for hardware intrinsics is in the JIT.
+
+### Hardware Intrinsics Table
+
+There is a hardware intrinsics table for each platform that supports hardware intrinsics: currently `_TARGET_XARCH_` and `TARGET_ARM64_`. They live in hwintrinsiclistxarch.h and hwintrinsiclistarm64.h respectively.
```

Comment:
It might be worth calling out that, for x86/x64; the intrinsics are shared between both and the x64 specific intrinsics are in a set of `X64` subclasses (e.g. `Sse` and `Sse.X64`). This was done because `x64` is a natural extension of `x86`.

For ARM/ARM64, these are currently considered separate micro-architectures; as it was raised the ARM64 ISA is not an extension of ARM32, but a new ISA altogether.
* The native intrinsics don't seem to care about this and just expose the same shared intrinsics, where applicable (generally referring to it as NEON)

Reply:
I would be interested in hearing more thoughts on this from the ARM team as they probably have the best insight as to whether we should have separate `System.Runtime.Intrinsics.Arm32` and `S.R.I.Arm64` namespaces or if we can/should allow some sharing of the intrinsics.

Reply:
I strongly advocated the separation of `System.Runtime.Intrinsics.Arm32` and `S.R.I.Arm64` namespaces. The main motivation was the per class `IsSupported` flag.  The HW flags indicating support for each extension are controlled by different HW registers. The set of instructions controlled by each flag was not the same. This resulted in the cleanest design IMHO.

Reply:
So ... the proposal would be to rename the `System.Runtime.Intrinsics.Arm.Arm64` namespace to `System.Runtime.Intrinsics.Arm64`? Seems reasonable to me, but that's a separate (API) discussion, and not really all that impactful to the JIT support, IMO.

Reply:
> So ... the proposal would be to rename the System.Runtime.Intrinsics.Arm.Arm64 namespace to System.Runtime.Intrinsics.Arm64? Seems reasonable to me, but that's a separate (API) discussion, and not really all that impactful to the JIT support, IMO.

I was mostly commenting on the current difference between x86/x64 and ARM32/ARM64 and indicating it might be good to elaborate on that here.

That is, when one platform is a natural extension of the other (i.e. `x64` is an extension of `x86`), we determined we should expose these centrally and have the public surface area and JIT support shared.

For ARM32/ARM64, where it is more of a redesign, rather than an extension; we determined that these should be separated even when some of the logic/instructions are still shared.
* To this point, I was also saying it would be good to get input from the ARM developers on whether they agree this is a good design. If it just comes down to `ARM32` uses `x` as the detection mechanism and `ARM64` uses `y`, but everything else is largely the same; then we might want to reconsider this for some ISAs.

Reply:
OK, so beyond the namespace issue, you're suggesting that, at such time as we support ARM32 intrinsics, we may want to reconsider whether there are opportunities for sharing in the JIT implementation?

Reply:
> The downside of this is though that user code is less portable as you'd need to use different namespaces when compiling for A32 and A64. In C/C++ we've called the headers the same for this reason, so the user can just recompile without any code changes.

There was a proposal to add a third namespace which were the common instructions supported by `Arm64` & `Arm32` to solve this problem.  That namespace would effectively be implemented in terms of the low-level platform specific ones. I don't think it was firmly agreed upon, but the concept was acknowledged as potentially feasible.

Reply:
> Though the problem would be how to share the namespace, If you call the top level one neon for instance which contains shared intrinsics between A32/A64 then you'd still need another one for the A64 only intrinsics if I understand that IsSupported can only be done at the class level.

The way it is done for x86 is we have the base "shared" ISA and then a `x64` specific subclass. So, for example:
```csharp
public abstract class Sse2 : Sse
{
    public bool IsSupported { get; }

    public abstract class X64 : Sse.X64
    {
        public bool IsSupported { get; }
    }
}
```

> If these can be solved in a clean way then I would advocate having a shared top level namespace mainly for portability between the two architectures.

I agree and if we determine we can't share these in the public surface area; that is probably fine. I just want to ensure that we put the same level of thought here as we did on the x86 side and with  input from the appropriate people; to ensure we ship something that is great and easy to use.

Reply:
So ... this is an interesting discussion but I think it's really an API design discussion.